File: atcoder.jp/Rust (rustc 1.70.0)/abc338/abc338_c.rs
use proconio::{fastout, input};

#[fastout]
fn main() {
    input! {
        n: usize,
        q: [i64; n],
        a: [i64; n],
        b: [i64; n],
    }

    let max = 10i64.pow(6);
    let mut ans = 0;
    for i in 0..=max {
        let rest = q
            .iter()
            .zip(a.iter())
            .map(|(q, a)| q - a * i)
            .collect::<Vec<_>>();
        if !rest.iter().all(|&r| r >= 0) {
            break;
        }
        let j = rest
            .iter()
            .zip(b.iter())
            .map(|(r, b)| if *b == 0 { max } else { r / b })
            .min()
            .unwrap();
        ans = ans.max(i + j);
    }
    println!("{}", ans);
}

File: atcoder.jp/Python (PyPy 3.10-v7.3.12)/abc338/abc338_c.py
N = int(input())
Q = list(map(int, input().split()))
A = list(map(int, input().split()))
B = list(map(int, input().split()))

inf = 1 << 60
ans = 0
for i in range(10**6 + 1):
    rest = [q - a * i for q, a in zip(Q, A)]
    if any([r < 0 for r in rest]):
        break
    j = min([r // b if b > 0 else inf for r, b in zip(rest, B)])
    ans = max(ans, i + j)

print(ans)
